showkase-component-impressions/src/main/java/com/vinaygaba/showkase_component_impressions/ComponentImpressions.kt

```diff
@@ -0,0 +1,145 @@
+    key: T,
+    onVisibilityChanged: (ShowkaseVisibilityEvent<T>) -> Unit,
+) = composed {
+    val view = LocalView.current
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
+    val scope = rememberCoroutineScope()
+    val visibilityEvent by rememberUpdatedState(newValue = onVisibilityChanged)
+    val impressionCollector = remember(key) { ImpressionCollector<T>(scope, visibilityEvent) }
+    var visibilityMetadata: VisibilityMetadata? by remember { mutableStateOf(null) }
+
+    registerLifecycleImpressionEvents(key, lifecycle, impressionCollector, visibilityMetadata)
+    registerDisposeImpressionEvents(key, impressionCollector, lifecycle)
+    collectImpressionEvents(key, impressionCollector, visibilityEvent)
+
+    onGloballyPositioned { layoutCoordinates ->
+        visibilityMetadata = layoutCoordinates.visibilityPercentage(view = view)
+        impressionCollector.onLayoutCoordinatesChanged(key, visibilityMetadata!!)
+    }
+}
+
+/**
+ * Used for handling the use case where a composable function is not in composition anymore i.e is
+ * invisible.
+ */
+@SuppressLint("ComposableNaming")
+@Composable
+private fun <T> registerDisposeImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    lifecycle: Lifecycle
+) {
+    DisposableEffect(key, lifecycle) {
+        onDispose {
+            impressionCollector.publishDisposeEvent(
+                key,
+                defaultVisibilityMetadata,
+            )
+        }
+    }
+}
+
+/**
+ * Collects the visibility change events from the internal coroutine flow and invokes the
+ * onVisibilityEvent that the user passed.
+ */
+@SuppressLint("ComposableNaming")
+@FlowPreview
+@Composable
+private fun <T> collectImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    onVisibilityEvent: (ShowkaseVisibilityEvent<T>) -> Unit
+) {
+    LaunchedEffect(key) {
+        impressionCollector.impressionEvents
+            .collect { impressionData ->
+                val impression = impressionData as ShowkaseVisibilityEvent<*>
+                onVisibilityEvent(
+                    ShowkaseVisibilityEvent(
+                        impression.key as T,
+                        impression.visibilityPercentage,
+                        impression.boundsInWindow
+                    )
+                )
+            }
+    }
+}
+
+/**
+ * Used for handling the use case where the activity/app is backgrounded or foregrounded.
+ */
+@SuppressLint("RememberReturnType", "ComposableNaming")
+@Composable
+private fun <T> registerLifecycleImpressionEvents(
+    key: T,
+    lifecycle: Lifecycle,
+    impressionCollector: ImpressionCollector<T>,
+    visibility: VisibilityMetadata?,
+) {
+    remember(key1 = key, key2 = visibility) {
+        lifecycle.addObserver(
+            LifecycleEventObserver { _, event ->
+                when (event) {
+                    Lifecycle.Event.ON_PAUSE, Lifecycle.Event.ON_STOP  -> {
```

Comment:
I don't think it makes sense to use both resume/start, as checking both isn't symmetric - ie you will publish event in start and then in pause. pick either resume/pause or start/stop to be symmetric

Reply:
i'm also not sure that it is ideal for the library to make the decision about which lifecycle events to use. some people might want created/destroyed, or they may not care about lifecycle changes and only want to know about percent visible.

maybe it would be better to add lifecycle state to the `ShowkaseVisibilityEvent` instead of internally using that information to set the visibility (in the spirit of being more generic and letting the user decide how they want to interpret the data)

the library can always add a util on top of that raw data as a default way to interpret it, eg an extension function like `ShowkaseVisibilityEvent.filterForLifecycle(Lifecycle.Event.ON_START)`